### ISSUE TITLE
Questions about adaptive scaling

### ISSUE THREAD

**Issue by lunzueta** [VSainteuf/metric-guided-prototypes-pytorch#3]
I wanted to use this approach in a project, but I had to reimplement it in TensorFlow instead of PyTorch, so I couldn't use this code directly. Everything worked fine except the adaptive scaling, which I had to put with a constant value of s=1. Otherwise, in the second epoch, the distortion loss would become NaN. I'm working on an image classification problem with data and a class hierarchy that I can't share yet. 

Did you find this kind of effect in any of your experiments (may after the publication)? I'm getting a bit crazy with it because everything looks fine in the code, and the results are also OK with s=1, the metric-guided approach works, but in a publication that we want to do about it, it looks a bit poor to say that we would fix s to 1. I'm not sure how to argue this... Do you think we might have something wrong with the code? Maybe TensorFlow does something internally that makes it go to NaN? Maybe it's because of our dataset and class hierarchy? Maybe it's not that "bad" to fix s in some cases? Any suggestions? Did you work further on this?

**Comment by VSainteuf** [comment]
Hey @lunzueta!
As far as I can remember I never had the scale free loss going NaN. It would be good to localise the problem more precisely. Like go in debug mode and execute the code line by line to see where the first NaN appears. 

From what you are saying it must appear somewhere in the ScaleFreeDistortion. 
So the main suspect would be `a_i` being too small, so that `scale=1/a_i` causes an overflow. 
If this is indeed the problem, clipping `a_i` during the first epochs would be a reasonable solution, as the scaling factor tends to decrease over the training procedure. 

Regarding putting `s=1`: Yes I would recommend fixing your nan problem, especially if you have a hierarchy with a lot of classes. We've seen (Tab. 3 of the Appendix) that the scale-free feature is especially useful for iNat19 that has over 1000 classes. But if your problem as fewer classes (10-100), the difference between scale-free and fixed-scale should be less important, or even non-existent (see results on CIFAR100 and NYU).

**Comment by lunzueta** [comment]
OK, thank you for your tips! We have 8 classes, so it's good to know what you say. I will try what you said about clipping during the first epochs.

**Comment by lunzueta** [comment]
I did the testing with the clipping, and I realized that in our case, with our data, the scale gets values much smaller than those you obtain in your experiments with the generalist datasets. I mean, we already start with values quite close to 0. Why could that be? Maybe because there is too much distortion between the prototypes and the target space?

**Comment by lunzueta** [comment]
Another thing I realized is that torch.norm might be different from tf.norm:
https://pytorch.org/docs/stable/generated/torch.norm.html
https://www.tensorflow.org/api_docs/python/tf/norm

By default, torch.norm uses the Frobenius norm, and tf.norm the Euclidean norm. But, in the paper, you talk about Euclidean distance. Could the type of norm be an important factor? Should your approach work in a similar way using different kinds of norms?

**Comment by VSainteuf** [comment]
I don't think the norm discrepancy is a problem. This is from the pytorch doc:

> Frobenius norm produces the same result as p=2 in all cases except when dim is a list of three or more dims, in which case Frobenius norm throws an error.

Regarding the scale, either the distance matrix takes smaller values or the distortion takes larger values than in our experiments. 
We defined the distance matrix as the length of the shortest path in the hierarchical tree so the values of D would be in the range of ~10 on average, are yours similar?
Then one thing that could make the distortion higher in your problem is if your embedding space has a much larger dimension. In our xps the maximum we tried was 512 for the embedding space of the iNat model. Is yours bigger than this ?

**Comment by lunzueta** [comment]
That's a good point because I'm using 1280 dimensions for the embedding space. I'll try with fewer dimensions, but what should be the limit, then?

**Comment by lunzueta** [comment]
I tried it, and although the scale took bigger numbers, it still failed with NaN values in the second iteration, so the problem was another one. In the end, I found [this](https://github.com/tensorflow/tensorflow/issues/12071#issuecomment-320797203) and manually defined the gradient of the norm function; now it trains correctly, even with 1280 dimensions. So, now I'll do some tests varying the dimensions, too, to see what gives the best results in my problem. Thanks!

**Comment by lunzueta** [comment]
FYI. Finally, the autoscaling won compared to setting manually different scales.

**Comment by VSainteuf** [comment]
Ah good to know thanks :) So at the end of the day, what was the problem in your first runs ?

**Comment by lunzueta** [comment]
It was what I said above about having to define the gradient of the norm function manually. So it was something internal from TensorFlow in the end.

**Comment by VSainteuf** [comment]
Ok thanks for the update!
